[front] - feat(skills): edit

JulesBelveze opened the pull request:
## Description

This PR implements the ability to edit existing skills in the skill builder. Previously, skills could only be created. Now users with proper authorization (skill editors or admins) can modify skill configurations including name, description, instructions, and tools.

The main changes include:
- Added a new page at `/w/[wId]/builder/skills/[sId]` for editing existing skills
- Implemented `PATCH` and `DELETE` API endpoints for skill configurations at `/api/w/[wId]/assistant/skill_configurations/[sId]`
- Added a new endpoint to fetch skill actions at `/api/w/[wId]/builder/skills/[sId]/actions`
- Enhanced `SkillConfigurationResource` with an `updateSkill` method and a `canEdit` property that checks if the current user has permission to edit
- Refactored `SkillBuilder` component to support both creation and editing modes
- Added navigation lock to prevent accidental data loss when editing
- Moved provider wrappers (`SpacesProvider`, `MCPServerViewsProvider`) to `SkillBuilderContext` for better organization
- Added reactive population of editors and tools from the server when editing an existing skill
- Removed the `scope` field from the skill builder form schema

## Risk

Breaking changes to skill configuration types: added `id` field to `SkillConfigurationType` which is now required in API responses. This could impact any code that depends on the existing type structure.

## Tests

- [x] Locally
- [ ] front-edge

## Deploy 

- Deploy front

JulesBelveze committed:
[front/api] - feature: implement skill configuration API endpoints

 - Added API endpoints to handle GET, PATCH, and DELETE operations for skill configurations
 - Integrated feature flag checks to conditionally allow access to skill-related features
 - Included validation for PATCH request body with `io-ts` to ensure data integrity
 - Implemented authorization checks to allow only editors or admins to modify or delete skill configurations
 - Added error handling to provide informative responses when operations fail or are unauthorized

JulesBelveze committed:
[front/pages/w/[wId]/builder/skills] - feature: add skill editing capability for users with proper authorization

 - Introduce a new page component allowing the editing of skills within a workspace
 - Implement server-side props fetching with user authentication and authorization checks
 - Ensure the feature is accessible only to users with editing rights and when the 'skills' feature flag is enabled
 - Include a redirect to a 404 page for unauthorized access or when the skill does not exist
 - Add layout integration and dynamic head title based on the skill being edited

JulesBelveze committed:
[skill_builder] - feature: add transform functions for skill form data

 - Implement function to convert server-side skill configurations to client-side form data
 - Provide a function to set default skill form data with user as editor for new skills
 - Add functionality for transforming a skill configuration into duplicate form data with modifications

JulesBelveze committed:
[front/lib/resources] - feature: enhance SkillConfigurationResource with new functionalities

 - Add method to fetch SkillConfiguration by its string ID with authorization
 - Implement canUserEdit method to check if a user has the right to edit a skill configuration
 - Introduce updateSkill method to handle updating skill details with version increment and re-fetching

JulesBelveze committed:
[front/pages/w/builder/skills] - feature: extend SkillsTable component with owner prop

 - SkillsTable now accepts an owner prop to possibly display owner-specific information or actions

JulesBelveze committed:
[skill_builder] - feature: enhance SkillBuilder to support editing and duplicating skills

 - Add the ability to edit existing skills and duplicate skill configurations
 - Implement navigation lock to prevent accidental data loss during creation or editing
 - Update form submission logic to handle both skill creation and updates depending on context
 - Improve form state management to reset editors and maintain form cleanliness on save
 - Adjust API calls to use PATCH method for updates and POST for new skill creation
 - Refactor SkillBuilder component to accept skill configuration and duplicate identifiers as props
 - Ensure that feedback notifications are context-aware, differing between creating and updating skills
 - Enable navigation to the skill editing page from the SkillsTable component

JulesBelveze committed:
[skill_builder] - refactor: streamline contexts and improve form data handling

- Remove MCPServerViewsProvider wrapper from SkillBuilderToolsSection
- Introduce useMCPServerViewsContext to handle server views within SkillBuilder
- Implement an effect to populate tools reactively based on MCP server views
- Adjust skillConfiguration types to SkillConfigurationType for consistency
- Include id field in SkillConfigurationType and related API responses
- Relocate provider wrappers to SkillBuilderProvider for better organization

JulesBelveze committed:
[skill_builder] - fix: resolve merge conflicts and cleanup code

 - Resolved git merge conflicts in multiple files including `SkillBuilder.tsx`, `submitSkillBuilderForm.ts`, `transformSkillConfiguration.ts`, and `SkillsTable.tsx`
 - Removed outdated or conflicting code related to the editing and duplication feature of skill configuration
 - Ensured that skill configuration updates and creation are handled correctly and consistent data is used in form handling
 - Refined skills table to use correct interaction for skill configuration actions, removing redundant router hooks and adapting to current logic
 - Added missing ID assignment for updated skill configuration in the API endpoint response

JulesBelveze committed:
[front] - refactor: remove skill duplication functionality from SkillBuilder

 - Eliminated code related to duplicating skills in `SkillBuilder` component
 - Removed the transformation function for duplicating skills' form data
 - Simplified form state reset logic by omitting checks for duplicate skill ID
 - Adjusted logic to ensure a more direct flow for skill creation or editing without duplication option

JulesBelveze committed:
[front] - refactor: remove id property from SkillsTable RowData

 - Eliminated the 'id' field from the row data structure to streamline the data model used in SkillsTable

JulesBelveze committed:
[front/lib/resources] - feature: add edit permission logic to SkillConfigurationResource

 - Implement canEdit property with logic to determine if a user can edit a skill
 - Optimize edit permission checks by using the new canEdit property across the application

[front/pages/api] - refactor: utilize canEdit property for edit permission checks

 - Replace direct permission checks with the new canEdit property for API endpoints
 - Simplify the logic by removing the redundant canUserEdit method

[front/pages] - refactor: update skill edit permission validation to use canEdit

 - Streamline server-side edit permission validation using the new canEdit property for skills
 - Remove unnecessary API calls for edit checks to improve performance and maintainability

JulesBelveze committed:
fix: lint/format

JulesBelveze committed:
[api/w/[wId]/builder/skills/[sId]] - feature: implement endpoint to get skill actions

 - Add new API endpoint to fetch actions configurations for a given skill
 - Ensure that only authenticated users with editor privileges or admins can access this endpoint
 - Validate that the feature 'skills' is enabled for the workspace before allowing access
 - Sanitize and format action names to be consistent with storage requirements

JulesBelveze committed:
fix: lint/format

fabiencelier commented on front/pages/api/w/[wId]/assistant/skill_configurations/[sId]/index.ts:
```diff
@@ -0,0 +1,293 @@
+        });
+      }
+
+      const updateResult = await skillResource.updateSkill(auth, {
```
Should we run everything in a transaction to avoid leaving it in a messy state if anytging fails in the middle ?

fabiencelier commented on front/lib/swr/skill_editors.ts:
```diff
@@ -0,0 +1,105 @@
+          body.removeEditorIds != null &&
+          body.removeEditorIds.length > 0
+        ) {
+          title = "Successfully update editors";
```
```suggestion
          title = "Successfully updated editors";
```

fabiencelier commented on front/components/skill_builder/submitSkillBuilderForm.ts:
```diff
@@ -1,81 +1,116 @@
-    if (editor.sId !== creatorId) {
-      addEditorIds.push(editor.sId);
+    if (!desiredEditorIds.has(creatorId)) {
+      removeEditorIds.push(creatorId);
```
I might be missing something here, but we ever only remove the "creatorId", not any other user

aubin-tchoi commented on front/lib/resources/skill_configuration_resource.ts:
```diff
@@ -256,6 +290,56 @@ export class SkillConfigurationResource extends BaseResource<SkillConfigurationM
+    try {
+      const workspace = auth.getNonNullableWorkspace();
+
+      await SkillConfigurationResource.model.update(
```
nit: this.model

aubin-tchoi reviewed (commented):
First pass, I stopped at front/lib/swr/actions.ts

JulesBelveze committed:
[skill_builder] - fix: correct skill editor management and update logic

 - Ensure that current editors are passed when submitting a skill builder form to manage editor differences correctly
 - Refactor transaction handling in skill update API to prevent inconsistent state during tool updates
 - Improve clarity in success messages when updating skill editors
 - Streamline the fetching of active skill configurations by name to avoid duplication
 - Optimize batch fetching of editor groups for skills to reduce database queries
 - Fix potential 404 error by ensuring proper parameter checks in the skill page server-side props

JulesBelveze committed:
[skill_builder] - refactor: improve error normalization and canEdit computation

- Standardize error handling in `submitSkillBuilderForm` using `normalizeError` utility function
- Extract `canEdit` logic into a separate method `computeCanEdit` for better readability and maintenance

aubin-tchoi reviewed (approved):
LGTM 🥵

JulesBelveze committed:
[front] - feature: add tests for skill configuration API endpoints

 - Implement Vitest test suites for GET, PATCH, and DELETE methods on skill configurations
 - Include validation tests for permissions, non-existent skills, and editor capabilities
 - Ensure transaction rollback is properly functioning on PATCH method when an error occurs
 - Verify that only supported HTTP methods are accepted, rejecting POST and PUT with appropriate error responses

JulesBelveze committed:
[skill_builder] - refactor: remove redundant 'user' parameter from skill form submission

 - 'user' parameter no longer needed when submitting skill builder forms, ensuring automatic editor addition is handled without explicit creator id
 - Simplify editor addition logic by removing unnecessary checks for the submitting user’s id

JulesBelveze committed:
```
[skill_builder] - refactor: update fetching and utilization of SkillConfigurationResource

 - Refactored methods in SkillConfigurationResource to reorder function definitions
 - Adjusted parameters in useSkillConfigurationTools to pass the entire owner object instead of just the ID
 - Enhanced API endpoint for skill configurations with validation for the skill ID parameter
```

JulesBelveze committed:
[front/lib/resources] - feature: implement updateTools method for SkillConfigurations

 - Add a new method to handle the updating of tool associations in SkillConfigurations
 - Ensure all tool associations are replaced transactionally with the provided MCP server view IDs

[front/pages/api] - refactor: streamline skill configuration update endpoint

 - Utilize the new updateTools method in skill configuration update logic
 - Remove direct manipulation of tool associations and use the new method to maintain consistency
 - Add validation for MCP server view IDs within the endpoint before processing the update transaction

JulesBelveze committed:
[front/components/skills] - refactor: change skill select mechanism in SkillsTable

- Replace router navigation with a setSkillConfiguration callback function for skill item clicks
- Remove useRouter dependency in SkillsTable component

[front/api] - fix: update SkillConfigurationResource fetching methods

- Change method used to fetch skill configurations from fetchById to fetchBySId for consistency and accuracy

JulesBelveze committed:
[front] - test: remove unnecessary skill creation in test

 - Eliminated the creation of an additional skill configuration not used in tests to streamline the testing process

JulesBelveze committed:
fix: lint/format

JulesBelveze committed:
[front] - test: ensure system space creation for MCP server in tests

 - Added steps to create a system space before creating an MCP server in API tests
 - This is required to fulfill preconditions for MCP server setup in the test environment

JulesBelveze committed:
test: stremaline tests
